pkg/engine/executor/range_aggregation.go

```diff
@@ -0,0 +1,287 @@
+}
+
+type rangeAggregationOptions struct {
+	partitionBy []physical.ColumnExpression
+
+	// start and end timestamps are equal for instant queries.
+	startTs       time.Time      // start timestamp of the query
+	endTs         time.Time      // end timestamp of the query
+	rangeInterval time.Duration  // range interval
+	step          *time.Duration // step used for range queries, nil for instant queries
+}
+
+// RangeAggregationPipeline is a pipeline that performs aggregations over a time window.
+// - It reads from the input pipelines
+// - Partitions the data by the specified columns
+// - Applies the aggregation function on each partition
+// Current version only supports counting for instant queries.
+type RangeAggregationPipeline struct {
+	state  state
+	inputs []Pipeline
+
+	aggregator *partitionAggregator
+	evaluator  *expressionEvaluator // used to evaluate column expressions
+	opts       rangeAggregationOptions
+}
+
+func NewRangeAggregationPipeline(inputs []Pipeline, evaluator *expressionEvaluator, opts rangeAggregationOptions) (*RangeAggregationPipeline, error) {
+	return &RangeAggregationPipeline{
+		inputs:     inputs,
+		evaluator:  evaluator,
+		aggregator: &partitionAggregator{entries: make(map[uint64]*partitionEntry)}, // TODO: estimate size during planning
+		opts:       opts,
+	}, nil
+}
+
+// Read reads the next value into its state.
+// It returns an error if reading fails or when the pipeline is exhausted. In this case, the function returns EOF.
+// The implementation must retain the returned error in its state and return it with subsequent Value() calls.
+func (r *RangeAggregationPipeline) Read() error {
+	// if the state already has an error, do not attempt to read.
+	if r.state.err != nil {
+		return r.state.err
+	}
+
+	if r.state.batch != nil {
+		r.state.batch.Release()
+	}
+
+	record, err := r.read()
+	r.state = newState(record, err)
+
+	if err != nil {
+		return fmt.Errorf("run range aggregation: %w", err)
+	}
+	return nil
+}
+
+// TODOs:
+// - Support implicit partitioning by all labels when partitionBy is empty
+// - Use columnar access pattern. Current approach is row-based which does not benefit from the storage format.
+// - Add toggle to return partial results on Read() call instead of returning only after exhausing all inputs.
+func (r *RangeAggregationPipeline) read() (arrow.Record, error) {
+	var (
+		isTSInRange  func(t time.Time) bool
+		tsColumnExpr = &physical.ColumnExpr{
+			Ref: types.ColumnRef{
+				Column: types.ColumnNameBuiltinTimestamp,
+				Type:   types.ColumnTypeBuiltin,
+			},
+		} // timestamp column expression
+
+		// reused on each row read
+		h           = xxhash.New()
+		labelValues = make([]string, len(r.opts.partitionBy))
+	)
+
+	{
+		evalTs := r.opts.endTs
+		earliestTs := r.opts.endTs.Add(-r.opts.rangeInterval)
+		isTSInRange = func(t time.Time) bool {
+			// Aggregate entries that belong in [earliestTs, evalTs)
+			return t.Compare(earliestTs) >= 0 && t.Compare(evalTs) < 0
+		}
```

Comment:
nit: You could create a helper struct like `bloomshipper.Interval` that has a `Compare(t time.Time)` method

Reply:
👍 we can add something like this when we add support for range queries